src/provider/provider_coarse.c

```diff
@@ -0,0 +1,1559 @@
+
+    block_t *merged_block = NULL;
+
+    umf_result =
+        block_merge_with_prev(upstream_memory_provider, all_blocks, free_blocks,
+                              block, false, &merged_block);
+    if (umf_result != UMF_RESULT_SUCCESS) {
+        // merge failed
+        return block;
+    }
+
+    // merge succeeded
+    assert(merged_block != NULL);
+
+    return merged_block;
+}
+
+// free_block_merge_with_next - merge the given free block
+// with the next one if both are unused and have continuous data.
+// Remove the merged block from the tree of free blocks.
+static block_t *free_block_merge_with_next(
+    umf_memory_provider_handle_t upstream_memory_provider,
+    struct ravl *all_blocks, struct ravl *free_blocks, block_t *block) {
+
+    assert(all_blocks);
+    assert(free_blocks);
+    assert(block);
+    assert(block->used == false);
+
+    // check if blocks could be merged by the upstream provider
+    umf_result_t merge_status = UMF_RESULT_ERROR_UNKNOWN;
+    if (block->next && block->next->used == false &&
+        (block->data + block->size == block->next->data)) {
+        if (upstream_memory_provider &&
+            (block->origin != block->next->origin)) {
+            merge_status = umfMemoryProviderAllocationMerge(
+                upstream_memory_provider, block->data, block->next->data,
+                block->size + block->next->size);
+        } else {
+            merge_status = UMF_RESULT_SUCCESS;
+        }
+    }
+
+    if (merge_status != UMF_RESULT_SUCCESS) {
+        return block;
+    }
+
+    block_t *to_free = block->next;
+
+    if (block->next->free_list_ptr) {
+        free_blocks_rm_node(free_blocks, block->next->free_list_ptr);
+        block->next->free_list_ptr = NULL;
+    }
+
+    assert(block->data < block->next->data);
+    assert((block->data + block->size) == block->next->data);
+
+    if (block->next->next) {
+        block->next->next->prev = block;
+    }
+
+    // set neighbors
+    block->size += block->next->size;
+    block->next = block->next->next;
+
+    block_t *block_rm = coarse_ravl_rm(all_blocks, to_free->data);
+    assert(block_rm == to_free);
+    (void)block_rm; // WA for unused variable error
+    umf_ba_global_free(to_free);
+
+    return block;
+}
+
+// alloc_find_origin - find the upstream allocation that contains data with given offset.
+static block_t *alloc_find_origin(block_t *alloc, size_t offset) {
+    assert(alloc);
+
+    while (offset >= alloc->size) {
+        offset -= alloc->size;
+        alloc = alloc->next;
+        assert(alloc);
+    }
+
+    return alloc;
+}
+
+typedef struct coarse_memory_provider_t {
+    umf_memory_provider_handle_t upstream_memory_provider;
+
+    // memory allocation strategy
+    coarse_memory_provider_strategy_t memory_strategy;
+
+    void *init_buffer;
+
+    size_t used_size;
+    size_t alloc_size;
+
+    // upstream_alloc - tree of all blocks allocated from the upstream provider
+    struct ravl *upstream_alloc;
```

Comment:
Not sure if we need this provider - we could free blocks from all_blocks ravl. We might need to call umfMemoryProviderSplit(), while spiting blocks (but TBH i do not know we are not doing this split)

Reply:
Yes, we do, because we have to free all blocks allocated from the upstream provider in the `finalize()` op. They are saved only in the `upstream_alloc` RAVL tree.

Reply:
Done